### PR TITLE
Make Shard Started Response Handling only Return after the CS Update Completes

### LINKED ISSUE #81628: Stop unnecessary retries of shard-started tasks
When a data node finishes recovering a shard it notifies the master to move it to state `STARTED`. Today we repeat this request every time we receive a cluster state that hasn't updated the shard state yet:

https://github.com/elastic/elasticsearch/blob/b6fbf5a1548c1924e67c360af2b6dd8ec51508ce/server/src/main/java/org/elasticsearch/indices/cluster/IndicesClusterStateService.java#L601-L619

This behaviour means if the master is busy processing (potentially thousands) of other `URGENT` tasks then we'll submit the same task repeatedly (potentially thousands of times). It dates back a long time but is no longer necessary: we can trust that the master will process our original request first (or we get notified that it failed). We should stop sending these unnecessary retries.

Relates https://github.com/elastic/elasticsearch/issues/77466

### PR DESCRIPTION
**Pull request by original-brownbear**
Somewhat lazy solution by copying the approach from the failed handler 1:1 for now.
Added a todo to clean up this thing.

closes #81628

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/cluster/action/shard/ShardStateAction.java [hunk 1]
```diff
@@ -614,15 +616,63 @@ public void messageReceived(StartedShardEntry request, TransportChannel channel,
...
                 shardStartedClusterStateTaskExecutor,
-                e -> {
-                    if (e instanceof FailedToCommitClusterStateException || e instanceof NotMasterException) {
-                        logger.debug(new ParameterizedMessage("failure during [shard-started {}]", request), e);
-                    } else {
-                        logger.error(new ParameterizedMessage("unexpected failure during [shard-started {}]", request), e);
+                new ClusterStateTaskListener() {
+                    @Override
+                    public void onFailure(Exception e) {
+                        logger.error(
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/cluster/action/shard/ShardStateAction.java:622]: I think it'd be better to keep the `DEBUG` level for `FailedToCommitClusterStateException` and `NotMasterException`. `ERROR` is a bit overdramatic in any case here.

**Review by idegtiarenko — APPROVED** [review]

File: server/src/main/java/org/elasticsearch/cluster/action/shard/ShardStateAction.java [hunk 2]
```diff
@@ -614,15 +616,63 @@ public void messageReceived(StartedShardEntry request, TransportChannel channel,
...
+                            () -> new ParameterizedMessage("{} unexpected failure while starting shard [{}]", request.shardId, request),
+                            e
+                        );
+                        try {
+                            channel.sendResponse(e);
+                        } catch (Exception channelException) {
+                            channelException.addSuppressed(e);
+                            logger.warn(
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/cluster/action/shard/ShardStateAction.java:630]: This is pretty much what `ChannelActionListener` does, maybe we should just use that?

File: server/src/main/java/org/elasticsearch/cluster/action/shard/ShardStateAction.java [hunk 3]
```diff
@@ -594,6 +595,7 @@ public void shardStarted(
         );
     }
 
+    // TODO: Make this a TransportMasterNodeAction and remove duplication of master failover retrying from upstream code
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/cluster/action/shard/ShardStateAction.java:598]: Mostly :+1: except that I believe `TransportMasterNodeAction` requires a timeout today but these things should not time out. Relates #82185 too I think.

**Review by DaveCTurner — COMMENTED** [review]
Looks good although I think we should have a test for it too. I left a few comments. I'll leave it to @idegtiarenko to review too as he's working in this area and might want to wait to avoid merge conflicts.

File: server/src/main/java/org/elasticsearch/cluster/action/shard/ShardStateAction.java (hunk 3 above)
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/cluster/action/shard/ShardStateAction.java:598]: Yea, but we can just do what we did for the snapshot shard state update and set it to the max value. We can even do nicer here and create an override of the master node request that doesn't physically write the always-max value redundantly.
I actually mostly implemented this already this morning but then figured this one I can get merged more quickly and it actually helps my benchmarks :)

File: server/src/main/java/org/elasticsearch/cluster/action/shard/ShardStateAction.java (hunk 2 above)
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/cluster/action/shard/ShardStateAction.java:630]: Yea right lets do that :)

File: server/src/main/java/org/elasticsearch/cluster/action/shard/ShardStateAction.java (hunk 1 above)
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/cluster/action/shard/ShardStateAction.java:622]: ++ made it debug in the `onNoLongerMaster` now.

Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/cluster/action/shard/ShardStateAction.java:628]: that fixes the `NotMasterException` case but not `FailedToCommitClusterStateException`

**Comment by original-brownbear** [comment]
>  I think we should have a test for it too.

Would be nice to have indeed. I don't see a quick way of adding one though. The exiting UT infrastructure doesn't seem to have the plumbing for this. For ITs (which I'd like better) I'd have to implement something along the lines of `org.elasticsearch.test.disruption.BusyMasterServiceDisruption` (but with actual state updates) and then use the mock transport to inspect whether or not I get a response before the task goes through or something along those lines?

Maybe again ok to leave this for later? :) I don't think I'll have the time to implement that today and don't want to block @idegtiarenko 's refactoring efforts here longer than necessary. I manually verified that this works correctly (seeing lots of dedup now actually happening in internal cluster tests that wouldn't happen before).

File: server/src/main/java/org/elasticsearch/cluster/action/shard/ShardStateAction.java (hunk 1 above)
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/cluster/action/shard/ShardStateAction.java]: :facepalm: right added conditional handling for `FailedToCommitClusterStateException` now as well.

**Review by DaveCTurner — APPROVED** [review]
LGTM

**Comment by original-brownbear** [comment]
Jenkins run elasticsearch-ci/part-1  (unrelated + known)

**Comment by original-brownbear** [comment]
Jenkins run elasticsearch-ci/part-1 (unrelated but new ... will open an issue)

**Comment by original-brownbear** [comment]
Thanks both!

**Pull request merged**